src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,279 @@
+      std::move(val),
+      [&element, levels, cb](struct_value& fields) {
+          if (fields.size() != element.children.size()) {
+              throw std::runtime_error(fmt::format(
+                "schema/struct mismatch, schema had {} children, struct had {} "
+                "fields. At column {}",
+                element.children.size(),
+                fields.size(),
+                element.column_index));
+          }
+          traversal_levels child_levels = levels;
+          // Increment the definition level as this node in the hierarchy is
+          // defined.
+          ++child_levels.definition_level;
+          for (size_t i = 0; i < element.children.size(); ++i) {
+              struct_field& member = fields[i];
+              const indexed_schema_element& child = element.children[i];
+              cb(child, child_levels, std::move(member.field));
+          }
+      },
+      [&element, levels, cb](null_value& v) {
+          // If the value is `null`, we use the parent definition_level so that
+          // assembly can determine where the `null` started.
+          for (const auto& child : element.children) {
+              cb(child, levels, value(v));
+          }
+      },
+      [&element](list_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected list value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](map_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected map value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](auto& v) {
+          throw std::runtime_error(fmt::format(
+            "unexpected leaf value for optional schema element {}: {}",
+            element.name,
+            value(std::move(v))));
+      });
+}
+
+void for_each_repeated_child(
+  value val,
+  const indexed_schema_element& element,
+  traversal_levels levels,
+  absl::FunctionRef<
+    void(const indexed_schema_element&, traversal_levels, value)> cb) {
+    ss::visit(
+      std::move(val),
+      [&element, levels, cb](null_value& v) {
+          for_each_optional_child(v, element, levels, cb);
+      },
+      [&element, levels, cb](list_value& list) {
+          // Empty lists are equivalent to a `null` value.
+          if (list.empty()) {
+              for_each_optional_child(null_value(), element, levels, cb);
+              return;
+          }
+          traversal_levels child_levels = levels;
+          // We are marking that there is a higher depth here we
+          // need to record, but the repetition_level is only set
+          // *when* we are repeating (so not for the first element).
+          ++child_levels.repetition_depth;
+          for (auto& member : list) {
+              for_each_optional_child(
+                std::move(member.element), element, child_levels, cb);
+              child_levels.repetition_level = child_levels.repetition_depth;
```

Comment:
nit: I was a little lost here.. guess the intention is to increment it only from the 2nd element on.. was wondering if we can do without this 3rd state repetition_depth.

Reply:
No we need the depth, because repeated could skip levels and you need to track which repeated level you're at when you write followup values. This is by far the simplest way I've seen of the different implementations.

Reply:
Alternatively the depth could be embedded in the schema, but I don't think that's much simpler TBH

Reply:
got it, I had to stare at the algorithm for a while but I think I got it now.